### ISSUE TITLE
Questions about project and wasm experience

### ISSUE THREAD

**Issue by scottbedard** [lgarron/cubing-rust-wasm#1]
Hi, I was wondering if you could share your experience with WASM so far, and it's learning curve. I'm working on a similar project, and am considering re-writing [this library](https://github.com/scottbedard/twister) with Rust for use on [speedcube.site](https://speedcube.site).

Currently, I'm just exporting the library as a node module for use by the frontend, and exposing a CLI for use by the backend. Has your experience with rust/wasm been good? It looks like this is in early development, so I understand if it's too soon to tell.

Thanks for your time! I'll definitely be following this repo 🙂

**Comment by lgarron** [comment]
Thanks for the interest!

At this point, the library is just an experiment. However, I'm fairly certain that there will be more and more of a need for portable high-performance cubing code. So far, Rust has been no less difficult to work with than TypeScript for https://github.com/cubing/cubing.js

Is there any particular functionality you need?

**Comment by scottbedard** [comment]
My needs are pretty specific to speedcube.site. In order to make keybindings as customizable as possible, I'm using some non-standard notation for higher order cubes and dodecaminxs. [This is the basic functionality I need](https://github.com/scottbedard/twister#apply)...

- Execute an algorithm
- Execute an algorithm backwards
- Test if the puzzle is in a solved state
- Serialize/deserialize puzzle state
- Generate a scramble algorithm
- Attach arbitrary data to stickers

My library may also be unique in that it stores state for each individual sticker, rather than each cubie. I chose this for gameplay rendering purposes, but I suspect it's not what most people would want. I'm not sure a shared library is in my best interest since I need to make changes as gameplay needs change.

I'm mostly interested in how the switch from TS to Rust has gone, especially when running in the context of a browser 🙂

**Comment by lgarron** [comment]
> I'm mostly interested in how the switch from TS to Rust has gone, especially when running in the context of a browser 🙂

This isn't a "switch", since the web still needs a library with a JS API. However, Rust is definitely faster for computational work, so we may consider using it inside `cubing.js` at some point (once bundlers properly handle WASM instantiation code).

**Comment by lgarron** [comment]
> * Execute an algorithm
> * Execute an algorithm backwards
> * Test if the puzzle is in a solved state
> * Serialize/deserialize puzzle state
> * Generate a scramble algorithm
> * Attach arbitrary data to stickers

`cubing.js` will easily handle those for you. 😉

It seems you're using some notation that is incompatible with wider community usage, though — I'd suggest reconsidering that if you still can, so that algs can be pasted between different apps.

**Comment by scottbedard** [comment]
Cubing.js is a great repo, and really that entire github org too! I considered using those libraries a while back, but decided it wasn't ideal to give up control of the state library since it's so tightly coupled to the needs of my gaming site.

I also considered using more standard notation, but existing ones I found were insufficient for my site's customization needs. I'm actually pretty happy with the notation I chose, but what I'm not happy with is performance as the puzzles scale. Anything larger than a gigaminx start to have a noticeable lag.

I think switching to Rust might help here, I just haven't had the pleasure of playing with WASM yet. It looks to be pretty mature now though, and Rust appears to have good support for it. So when the name of this repo showed up on my feed, it immediately caught my attention 🙂
